manually-adding-services-to-metrics.adoc

```diff
@@ -0,0 +1,120 @@
+= Manually making a service visible to the Metrics Service
+
+== Prerequesites
+
+* OpenShift Origin >= 3.7
+** https://docs.openshift.org/latest/install_config/index.html
+* OpenShift Ansible Broker
+** https://docs.openshift.org/latest/install_config/install/advanced_install.html#configuring-openshift-ansible-broker
+* Metrics service deployment from the Service Catalog
+** https://github.com/aerogearcatalog/metrics-apb
+
+== Adding a service to the Metrics Service
+
+The Metrics Service is composed of both https://grafana.com/[Grafana]
+for data visualization and https://prometheus.io/[Prometheus] for
+gathering of metrics data, so including a new service involves
+configuring the Prometheus instance to include scraping of metrics
+endpoints for the new service.
+
+=== Service requirements
+
+In order to provide data that is consumable (scrape-able) by Prometheus,
+the service must expose a *metrics endpoint*, see the
+https://prometheus.io/docs/instrumenting[Prometheus documentation on
+instrumenting] your own services to add metrics endpoints to an existing
+application.
+
+The endpoint must:
+
+* Not have any authentication requirements
+* Be visible by the Prometheus instance
+
+=== Adding a new scraping configuration to the Prometheus instance
+
+In this example we will edit the existing Prometheus configuration to
+include a new `scrape_config` targetting our service.
+
+This example will target the Grafana instance deployed alongside
+Prometheus, which exposes a metrics endpoint in the default `/metrics`
+HTTP endpoint.
+
+See the
+link:https://prometheus.io/docs/prometheus/latest/configuration/configuration/\#scrape_config[Prometheus
+documentation on scrape_config] for more details.
+
+=== Via the OpenShift Console
+
+1.  Select your project in the `My Projects` sidebar
+
+image:img/manually-adding-services-to-metrics/myprojects.png[My
+projects]
+
+1.  Navigate to Prometheus' Config Map via
+`Resources > Config Maps > prometheus-configmap`
+
+image:img/manually-adding-services-to-metrics/configmap.png[Config Map]
+
+1.  Enter editing mode through `Actions > Edit YAML`
+
+image:img/manually-adding-services-to-metrics/edit.png[Edit YAML]
+
+1.  Include a new `scrape_config` targetting the Grafana service:
+
+[source,yaml]
+----
+apiVersion: v1
+data:
+  prometheus.yml: |-
+    ---
+    global:
+      evaluation_interval: 5s
+      scrape_interval: 5s
+    scrape_configs:
+      - job_name: prometheus
+        static_configs:
+          - targets:
+            - "localhost:9090"
+      - job_name: grafana
+        scheme: https
+        static_configs:
+          - targets:
+            - "grafana.myproject.svc:4181"
```

Comment:
@david-martin I'm having trouble make the grafana endpoint actually work because it's behind the openshift auth proxy.

I was wondering if it would be relevant to actually use @aliok 's [sample node app](https://github.com/aliok/node-app-with-prometheus-metrics) instead and if we should be looking into exposing the grafana metrics endpoint without auth.

IIUC we created the prometheus-internal service to enable grafana to reach it without going through the proxy, but is there a way to configure the proxy to let requests to those specific endpoints (/metrics and whatever grafana has to query prometheus for datases) through?

cc @darahayes

Reply:
@paolobueno If my understanding is correct this document is showing an example where prometheus is scraping metrics out of the existing Grafana service is that correct?

Perhaps there is a way to allow specific requests through the proxy without auth. Whether it's based on endpoint or origin or something else. However I don't see this as the best approach. In this scenario Prometheus would be talking to Grafana through its route as defined in OpenShift.

It's my understanding that OpenShift routes are used to publicly expose services on the internet in a production scenario. I may be wrong but I think the convention in Kubernetes/Openshift is that applications in the same namespace should be talking to each other using a Service. Similar to the prometheus-internal service.

Reply:
@darahayes Yes, I've only picked grafana as the example target in the context of the guide because 1. it's not currently labelled for discovery by prometheus, so manual addition would be required and 2. it's already provisioned alongside it, so there wouldn't be a prereq for another apb.

Ah yes, the 4181 port there is wrong, forgot to update. The grafana service exposes the default https 443 instead, while 4181 is indeed internal to the pod.

I won't update it right now not to flag this thread as out-of-date, but my take is that this should be:

```yaml
- job_name: grafana
  scheme: https
  tls_config:
    insecure_skip_verify: true
  static_configs:
    - targets:
      - "grafana:443"
```

Acessing grafana either through the service binding or even from the public route (https://grafana-myproject.192.168.37.1.nip.io/metrics) still go through the proxy and the response is 403 since the prometheus client isn't authenticated, so that's where I'm stuck.

Reply:
@paolobueno You can use thie `-skip-auth-regex` flag in the oauth proxy
```
value: bypass authentication for requests path's that match (may be given multiple times). Cannot be set with -bypass-auth-except-for
```
https://github.com/openshift/oauth-proxy#command-line-options

Reply:
@paolobueno yes the two options are:

1. Prometheus talks to Grafana through the oauth proxy. This means figuring out how to allow the requests through the proxy. It looks like @david-martin's suggestion might help with that.

2. Create a new service that directly exposes Grafana within the namespace.

There are benefits and drawbacks to each approach I guess.

Reply:
Thanks a lot for the help. I guess creating an internal service for grafana too might be the best way to go, since exposing the metrics endpoints to the outside unconditionally might have security implications... going to move forward on this for now.

Reply:
+1 for providinig an example using an internal service to access to grafana metrics endpoint